File: 2. 两数相加/rust/main/src/main.rs
// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
  pub val: i32,
  pub next: Option<Box<ListNode>>
}

impl ListNode {
  #[inline]
  fn new(val: i32) -> Self {
    ListNode {
      next: None,
      val
    }
  }
}

struct Solution;

impl Solution {
    pub fn add_two_numbers(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut head = Some(Box::new(ListNode::new(0)));
        let mut p = &mut head;
        let mut p1 = &l1;
        let mut p2 = &l2;
        let mut add1 = false;

        while *p1 != None || *p2 != None || add1 {
            let mut num = Self::get_val(p1) + Self::get_val(p2);
            if add1 {
                num += 1;
            }
            add1 = num >= 10;
            p.as_mut().unwrap().next = Some(Box::new(ListNode::new(num % 10)));
            p = &mut p.as_mut().unwrap().next;
            p1 = match p1 {
                Some(n) => &n.next,
                None => &None,
            };
            p2 = match p2 {
                Some(n) => &n.next,
                None => &None,
            };
        }
        head.unwrap().next
    }

    fn get_val(node: &Option<Box<ListNode>>) -> i32 {
        match node {
            None => 0,
            Some(n) => {
                n.val
            }
        }
    }
}

fn main() {
    println!("Hello, world!");
}

File: 2. 两数相加/swift/main/main/main.swift
//
//  main.swift
//  main
//
//  Created by cedricwu on 7/21/25.
//

import Foundation


public class ListNode {
    public var val: Int
    public var next: ListNode?
    public init() { self.val = 0; self.next = nil; }
    public init(_ val: Int) { self.val = val; self.next = nil; }
    public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
}

class Solution {
    func addTwoNumbers(_ l1: ListNode?, _ l2: ListNode?) -> ListNode? {
        var head = ListNode()
        var p: ListNode? = head
        var add1 = false
        
        var p1 = l1
        var p2 = l2
        while p1 != nil || p2 != nil || add1 {
            var num = (p1?.val ?? 0) + (p2?.val ?? 0) + (add1 ? 1 : 0)
            p?.next = ListNode(num % 10)
            
            add1 = num >= 10
            p = p?.next
            p1 = p1?.next
            p2 = p2?.next
        }
        
        return head.next
    }
}
